pkg/collector/corechecks/servicediscovery/envs/envs.go

```diff
@@ -0,0 +1,86 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+// Package envs provides target environment variables of interest.
+package envs
+
+import (
+	"fmt"
+)
+
+// targets is a collection of environment variables of interest.
+var targets = map[string]struct{}{
+	"PWD":                      {},
+	"DD_INJECTION_ENABLED":     {},
+	"DD_SERVICE":               {},
+	"DD_TAGS":                  {},
+	"DD_DISCOVERY_ENABLED":     {},
+	"GUNICORN_CMD_ARGS":        {},
+	"WSGI_APP":                 {},
+	"CORECLR_ENABLE_PROFILING": {},
+	"CATALINA_OPTS":            {},
+	"JAVA_TOOL_OPTIONS":        {},
+	"_JAVA_OPTIONS":            {},
+	"JDK_JAVA_OPTIONS":         {},
+	"JAVA_OPTIONS":             {},
+	"JDPA_OPTS":                {},
+	"SPRING_APPLICATION_NAME":  {},
+	"SPRING_CONFIG_LOCATIONS":  {},
+	"SPRING_CONFIG_NAME":       {},
+	"SPRING_PROFILES_ACTIVE":   {},
+}
+
+// GetExpectedEnvs - return list of expected env. variables for testing.
+func GetExpectedEnvs() ([]string, map[string]string) {
+	var expectedEnvs []string
+	var expectedMap = make(map[string]string)
+
+	for env := range targets {
+		expectedEnvs = append(expectedEnvs, fmt.Sprintf("%s=true", env))
+		expectedMap[env] = "true"
+	}
+	return expectedEnvs, expectedMap
+}
+
+// EnvironmentVariables - collected of targeted environment variables.
+type EnvironmentVariables struct {
+	vars map[string]string
+}
+
+// NewEnvironmentVariables returns a new [EnvironmentVariables] to collect env. variables.
+func NewEnvironmentVariables(vars map[string]string) EnvironmentVariables {
+	return EnvironmentVariables{
+		vars: vars,
+	}
+}
+
+// GetVars returns the collected environment variables
+func (ev *EnvironmentVariables) GetVars() map[string]string {
+	return ev.vars
+}
+
+// Get returns an environment variable if it is present in the collection
+func (ev *EnvironmentVariables) Get(name string) (string, bool) {
+	if _, ok := targets[name]; !ok {
+		return "", false
+	}
+
+	val, ok := ev.vars[name]
+	return val, ok
```

Comment:
imho, `vars` should reflect only the allowed environment variables
no need to store (and waste memory) unneeded envs + currently you always pay for 2 map accesses per "supported" environment variable instead of only one map access

Reply:
If `vars` is guaranteed to have only supported environment variables, then no need to check `targets` first

Reply:
This is where unnecessary environment variables are filtered out and ultimately only the requested variables are included in the list.

Reply:
I don't get the code then, and how you spare redundant memory

We have 100 environment variable for process X, we care about only 10 of them
We store the entire 100 environment variables in the memory, but we only allow accessing the 10?

The filtering should be while reading the envs or when creating the `Variables`
We should not read redundant environment variable, we should not store redundant environment variable, and we should not allow the consumers of `envs.Variables` to access redundant environment variables

As far as I understand the code, only the third requirement is fulfilled.
We should at least fulfill the second requirement as well

Reply:
The Set() command is called while reading file. Flow is Scan() -> returns next string -> add() -> Set() -> check if need it -> Scan(next). Only target variables are stored.

Reply:
Expanding my previous answer. The file scanner reads "environ" file string by string and returns true when next string was read. The "add" function fetches current string and passes env variable name and value to envs.Set, this function saves only targeted variable and scanner goes to the next string in file. If variable is not of interest it is discarded and not stored in any place.

Reply:
What is perhaps not very clear is that the check on `Get` is for test cases. In the test cases the map passed in to `NewVariables` could have variables which are not on the allowlist (in case new code is being added), so for this check on `Get` prevents these variables from being used until they have been added to the `allowlist`, to avoid missing new variables. The check on `Get` could be dropped if the `NewVariables` function is made to do the check for valid variables instead. This will only affect the tests, since in the normal case nil or an empty mapped will be passed in to NewVariables() since variables are added as they are read.

Reply:
I'd suggest the above change to move the check to NewVariables be done in a follow-up PR.

Reply:
Ok, I'll modify Get() function and will move target verification to stage of struct initialization in a follow-up PR

Reply:
this is follow-up [PR ](https://github.com/DataDog/datadog-agent/pull/30068)